server/src/main/java/org/elasticsearch/snapshots/SnapshotsService.java

```diff
@@ -1887,55 +1887,67 @@ public boolean assertAllListenersResolved() {
+                        logger.trace("[{}] Updating shard [{}] with status [{}]", updatedSnapshot,
                                 finishedShardId, updateSnapshotState.status().state());
-                        if (updated == false) {
-                            shards.putAll(entry.shards());
-                            updated = true;
+                        if (shards == null) {
+                            shards = ImmutableOpenMap.builder(entry.shards());
                         }
                         shards.put(finishedShardId, updateSnapshotState.status());
                         changedCount++;
                     } else {
-                        final String updatedRepository = updateSnapshotState.snapshot().getRepository();
-                        final Set<ShardId> reusedShardIds = reusedShardIdsByRepo.computeIfAbsent(updatedRepository, k -> new HashSet<>());
-                        if (entry.state().completed() == false && entry.repository().equals(updatedRepository)
-                                && reusedShardIds.contains(finishedShardId) == false) {
-                            final ShardSnapshotStatus existingStatus = entry.shards().get(finishedShardId);
-                            if (existingStatus == null || existingStatus.state() != ShardState.QUEUED) {
-                                continue;
-                            }
-                            if (updated == false) {
-                                shards.putAll(entry.shards());
-                                updated = true;
-                            }
-                            final ShardSnapshotStatus finishedStatus = updateSnapshotState.status();
-                            logger.trace("Starting [{}] on [{}] with generation [{}]", finishedShardId,
-                                    finishedStatus.nodeId(), finishedStatus.generation());
-                            shards.put(finishedShardId, new ShardSnapshotStatus(finishedStatus.nodeId(), finishedStatus.generation()));
-                            reusedShardIds.add(finishedShardId);
+                        final ShardSnapshotStatus existingStatus = entry.shards().get(finishedShardId);
+                        if (existingStatus == null || existingStatus.state() != ShardState.QUEUED) {
+                            continue;
+                        }
+                        if (shards == null) {
+                            shards = ImmutableOpenMap.builder(entry.shards());
                         }
+                        final ShardSnapshotStatus finishedStatus = updateSnapshotState.status();
+                        logger.trace("Starting [{}] on [{}] with generation [{}]", finishedShardId,
+                                finishedStatus.nodeId(), finishedStatus.generation());
+                        shards.put(finishedShardId, new ShardSnapshotStatus(finishedStatus.nodeId(), finishedStatus.generation()));
+                        iterator.remove();
```

Comment:
`changedCount++;`?

Reply:
We only counted the updates to existing running snapshots so far but not the updates that made new shard snapshots start. I don't have much of an opinion here but I think I like the current approach marginally better than counting up here as well. It at least allows you to judge how well state update batching is working if you turn on trace logging (if you count an update either once or twice it's more confusing).

Reply:
Note that we later on down use `if (changedCount)` to determine if a CS update is needed.

Reply:
Perhaps we can count both separately (and  expose in log message)

Reply:
> Note that we later on down use if (changedCount) to determine if a CS update is needed.

It is actually very important that we do that. Otherwise a dangling data node that re-sends its shard updates long after the snapshot has finished might result in two queued up snapshots having the same shard start snapshotting.

I just realized that even with this check that is still a possibility though and we have a (very unlikely) bug here in fact in situations where data nodes resend their updates after a snapshot has finished.

Reproducing this requires a pretty tricky test and a fix needs additional changes to this loop. Maybe we can take the straight forward speedups that don't change behavior in here here separately now as is and I'll refactor this in a follow-up that also fixes the above bug?

Reply:
Nevermind actually, the fix fits in here just fine. On it.

Reply:
Alright I pushed https://github.com/elastic/elasticsearch/pull/62070/commits/5c05dc11c8a626f625d7833a58ea5194368aa12b and https://github.com/elastic/elasticsearch/pull/62070/commits/eb14918fc193edaefb8e302915fdcb6c65fd7bea to keep two counts for the log message.

Also, I made a change there that fixes the bug I was talking about. Currently, if a dangling data node were to send a shard update for a snapshot that is long gone from the CS it might trigger a queued up shard snapshot to start and another snapshot for that shard is already running. This is only a possibility if it gets batched up with a shard update for an existing in-progress snapshot due to the `changedCount` condition so it's incredibly unlikely in practice and I'm having a very hard timing coding up a test for it (even in `SnapshotsResiliencyTests` I can't seem to find a scenario where this kind of CS update batching occurs at any non-trivial frequency so far). I still think this is a valid fix, plus it actually improves performance to check if a task has even been reused as well to a trivial degree in some corner cases I suppose.